arbnode/blockmetadata.go

```diff
@@ -0,0 +1,167 @@
+	var err error
+	var jwt *common.Hash
+	if c.JWTSecret != "" {
+		jwt, err = signature.LoadSigningKey(c.JWTSecret)
+		if err != nil {
+			return nil, fmt.Errorf("BlockMetadataRebuilder: error loading jwt secret: %w", err)
+		}
+	}
+	var client *rpc.Client
+	if jwt == nil {
+		client, err = rpc.DialOptions(ctx, c.Url)
+	} else {
+		client, err = rpc.DialOptions(ctx, c.Url, rpc.WithHTTPAuth(node.NewJWTAuth([32]byte(*jwt))))
+	}
+	if err != nil {
+		return nil, fmt.Errorf("BlockMetadataRebuilder: error connecting to bulk blockMetadata API: %w", err)
+	}
+	return &BlockMetadataRebuilder{
+		config: c,
+		db:     db,
+		client: client,
+		exec:   exec,
+	}, nil
+}
+
+func (b *BlockMetadataRebuilder) Fetch(ctx context.Context, fromBlock, toBlock uint64) ([]gethexec.NumberAndBlockMetadata, error) {
+	var result []gethexec.NumberAndBlockMetadata
+	err := b.client.CallContext(ctx, &result, "arb_getRawBlockMetadata", rpc.BlockNumber(fromBlock), rpc.BlockNumber(toBlock))
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func ArrayToMap[T comparable](arr []T) map[T]struct{} {
+	ret := make(map[T]struct{})
+	for _, elem := range arr {
+		ret[elem] = struct{}{}
+	}
+	return ret
+}
+
+func (b *BlockMetadataRebuilder) PushBlockMetadataToDB(query []uint64, result []gethexec.NumberAndBlockMetadata) error {
+	batch := b.db.NewBatch()
+	queryMap := ArrayToMap(query)
+	for _, elem := range result {
+		pos, err := b.exec.BlockNumberToMessageIndex(elem.BlockNumber)
+		if err != nil {
+			return err
+		}
+		if _, ok := queryMap[uint64(pos)]; ok {
+			if err := batch.Put(dbKey(blockMetadataInputFeedPrefix, uint64(pos)), elem.RawMetadata); err != nil {
+				return err
+			}
+			if err := batch.Delete(dbKey(missingBlockMetadataInputFeedPrefix, uint64(pos))); err != nil {
+				return err
+			}
+		}
+	}
+	return batch.Write()
+}
+
+func (b *BlockMetadataRebuilder) Update(ctx context.Context) time.Duration {
+	handleQuery := func(query []uint64) bool {
+		result, err := b.Fetch(
+			ctx,
+			b.exec.MessageIndexToBlockNumber(arbutil.MessageIndex(query[0])),
+			b.exec.MessageIndexToBlockNumber(arbutil.MessageIndex(query[len(query)-1])),
+		)
+		if err != nil {
+			log.Error("Error getting result from bulk blockMetadata API", "err", err)
+			return false
+		}
+		if err = b.PushBlockMetadataToDB(query, result); err != nil {
+			log.Error("Error committing result from bulk blockMetadata API to ArbDB", "err", err)
+			return false
+		}
+		return true
+	}
+	iter := b.db.NewIterator(missingBlockMetadataInputFeedPrefix, nil)
+	defer iter.Release()
+	var query []uint64
+	for iter.Next() {
+		keyBytes := bytes.TrimPrefix(iter.Key(), missingBlockMetadataInputFeedPrefix)
+		query = append(query, binary.BigEndian.Uint64(keyBytes))
+		end := len(query) - 1
+		if query[end]-query[0]+1 >= uint64(b.config.APIBlocksLimit) {
+			if query[end]-query[0]+1 > uint64(b.config.APIBlocksLimit) && len(query) >= 2 {
```

Comment:
Yep, we expect them to fragmented whenever a sequencer goes-offline and comes backup and syncs via a source that can't provide BlockMetadata- like a Parent chain. But we do expect they'll be in a set of groups like- 
* 1-100 have blockMetadata, 
* 101-150 doesn't and, 
* 151-200 have 

Gaps frequency in case of sequencer nodes shouldn't be high (as they sync via redis) and for the normal nodes it will depend on the feed sources it is connected to, but I think it will be on the lower side too.

> Just wondering if we should bother with inefficiency from requesting metadata for more blocks than we need (e.g. if we need only metadata for few blocks from range) - I guess it shouldn't be the case.

We do have a solution in place for this (by sacrificing the number of times we call the API) by setting `APIBlocksLimit` to a reasonable number, this ensures we dont fetch a lot of unnecessary data, For example blocks 20-30 are missing and then from block 200-210, by setting the `APIBlocksLimit` to say 50, we avoid fetching 20 to 210 instead we do two fetches 20-30 and 200-210.

Reply:
added this